app/src/main/java/bg/zahov/app/ui/workout/WorkoutViewModelNew.kt

```diff
@@ -0,0 +1,905 @@
+     * @property setCompleted Indicates whether the set has been completed. Default is false.
+     */
+    data class SetEntry(
+        val id: String = generateRandomId(),
+        val setType: SetType = SetType.DEFAULT,
+        val firstInputFieldVisibility: Boolean = true,
+        val setNumber: String,
+        val previousResults: String = "-//-",
+        val set: Sets,
+        val setCompleted: Boolean = false,
+    ) : WorkoutEntry()
+}
+
+/**
+ * ViewModel responsible for managing the state and logic of a workout session.
+ *
+ * @property workoutStateManager Manages the state of the workout session.
+ * @property repo Provides workout data and operations related to workouts.
+ * @property addExerciseToWorkoutProvider Handles adding exercises to the current workout session.
+ * @property restTimerProvider Manages timers for rest periods during the workout.
+ * @property settingsProvider Provides access to user settings, such as units and preferences.
+ * @property toastManager Manages the display of toast messages for user feedback.
+ */
+class WorkoutViewModelNew(
+    private val workoutStateManager: WorkoutStateManager = Inject.workoutState,
+    private val repo: WorkoutProvider = Inject.workoutProvider,
+    private val addExerciseToWorkoutProvider: AddExerciseToWorkoutProvider = Inject.workoutAddedExerciseProvider,
+    private val restTimerProvider: RestTimerProvider = Inject.restTimerProvider,
+    private val settingsProvider: SettingsProvider = Inject.settingsProvider,
+    private val toastManager: ToastManager = ToastManager,
+) : ViewModel() {
+
+    /**
+     * A private mutable state flow that represents the UI state of the workout.
+     */
+    private val _uiState = MutableStateFlow<WorkoutUiState>(WorkoutUiState())
+
+    /**
+     * A public read-only state flow exposing the current workout UI state to observers.
+     */
+    val uiState: StateFlow<WorkoutUiState> = _uiState
+
+    /**
+     * The index of the exercise to replace in the current workout, if applicable.
+     */
+    private var exerciseToReplaceIndex: Int? = null
+
+    /**
+     * A list of template exercises used to lookup previous results
+     */
+    private var templateExercises = listOf<Exercise>()
+
+    /**
+     * Units used by the user
+     */
+    private lateinit var units: Units
+
+    /**
+     * Received by arguments. The id of the template workout of which
+     * this workout will be based on
+     */
+    private var workoutId: String? = null
+
+    /**
+     * The date and time when the workout session was created.
+     */
+    private val workoutDate: LocalDateTime = LocalDateTime.now()
+
+    /**
+     * Initializes the `WorkoutViewModel` by setting up multiple asynchronous data flows to manage the UI state
+     */
+    init {
+        viewModelScope.launch {
+            // Observes whether the workout state should be saved and triggers the save operation.
+            launch {
```

Comment:
Extract every process into separate method for better readability and maintenance. You can  use viewModelScope.launch in every sub function

Reply:
🫡